stl/inc/algorithm

```diff
@@ -3285,13 +3310,234 @@ _FwdIt2 remove_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _P
 template <class _ExPo, class _FwdIt, class _Ty, _Enable_if_execution_policy_t<_ExPo> = 0>
 _NODISCARD _FwdIt remove(
     _ExPo&& _Exec, const _FwdIt _First, const _FwdIt _Last, const _Ty& _Val) noexcept; // terminates
-#endif // _HAS_CXX17
 
-#if _HAS_CXX17
 template <class _ExPo, class _FwdIt, class _Pr, _Enable_if_execution_policy_t<_ExPo> = 0>
 _NODISCARD _FwdIt remove_if(_ExPo&& _Exec, _FwdIt _First, const _FwdIt _Last, _Pr _Pred) noexcept; // terminates
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::remove
+    class _Remove_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <permutable _It, sentinel_for<_It> _Se, class _Ty, class _Pj = identity>
+            requires indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>
+        constexpr subrange<_It> operator()(_It _First, _Se _Last, const _Ty& _Val, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _ULast = _Get_unwrapped(_STD move(_Last));
+            auto _UFirst  = _Find_unchecked(_Get_unwrapped(_STD move(_First)), _ULast, _Val, _Pass_fn(_Proj));
+            auto _UResult = _Remove_unchecked(_UFirst, _ULast, _Val, _Pass_fn(_Proj));
+
+            return _Rewrap_subrange<subrange<_It>>(_First, _STD move(_UResult));
+        }
+
+        template <forward_range _Rng, class _Ty, class _Pj = identity>
+            requires permutable<iterator_t<_Rng>>
+                  && indirect_binary_predicate<ranges::equal_to, projected<iterator_t<_Rng>, _Pj>, const _Ty*>
+        constexpr borrowed_subrange_t<_Rng> operator()(_Rng&& _Range, const _Ty& _Val, _Pj _Proj = {}) const {
+            auto _ULast   = _Uend(_Range);
+            auto _UFirst  = _Find_unchecked(_Ubegin(_Range), _ULast, _Val, _Pass_fn(_Proj));
+            auto _UResult = _Remove_unchecked(_UFirst, _ULast, _Val, _Pass_fn(_Proj));
+
+            return _Rewrap_subrange<borrowed_subrange_t<_Rng>>(_Range, _STD move(_UResult));
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Ty, class _Pj>
+        _NODISCARD static constexpr subrange<_It> _Remove_unchecked(
+            _It _First, const _Se _Last, const _Ty& _Val, _Pj _Proj) {
+            // Remove projected values equal to _Val from [_First, _Last)
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>);
+
+            auto _Next = _First;
+            for (; _First != _Last; ++_First) {
+                if (_STD invoke(_Proj, *_First) != _Val) {
+                    *_Next = _RANGES iter_move(_First);
+                    ++_Next;
+                }
+            }

```

Comment:
This potentially destroys every element in the range before the first match by self-move-assignment.
```suggestion
            _First = _RANGES _Find_unchecked(_STD move(_First), _Last, _Val, _Proj);
            auto _Next = _First;
            if (_First != _Last) {
                while (++_First != _Last) {
                    if (_STD invoke(_Proj, *_First) != _Val) {
                        *_Next = _RANGES iter_move(_First);
                        ++_Next;
                    }
                }
            }
```

Reply:
By construction that cannot happen. Note that `_First` which is passed into `_Remove_unchecked` is the already return value of `_Find_unchecked` (line 3342).

I could move the check here, but I found the implementation as it is clearer like this

Reply:
> Note that `_First` which is passed into `_Remove_unchecked` is the already return value of `_Find_unchecked` (line 3345).

Derp 🤦 . I'll take your word for this being clearer despite the duplicated code in the callers, but we should name this something other than `_Remove_unchecked` which suggests that it's a complete implementation of the algorithm.

Reply:
Thanks so much for bringing this up. I have been shouting consistency all the time now and this clearly fails at it.

To add insult to injury this is only simpler because it is incorrect in that it would compare the first -to-be-removed- element twice. Once during find and once during remove. I will update the code to use the original algorithm